hackage-security/src/Hackage/Security/Client.hs

```diff
@@ -335,6 +338,8 @@ cachedVersion CachedInfo{..} remoteFile =
 getCachedInfo ::
 #if __GLASGOW_HASKELL__ < 800
                  (Applicative m, MonadIO m)
+#elif __GLASGOW_HASKELL__ >= 807
+                 (MonadFail ReadJSON_Keys_Layout, MonadIO m)
```

Comment:
Indeed this seems rather odd. I haven't looked at the surrounding code but it does seem suspicious that you would need to add a constraint of this form. Are you sure this is necessary?

Reply:
I'm pretty sure I did it all wrong, I'll dig into it tomorrow.

Reply:
I'm not clear how `MonadFail` can be replaced, it seems to come from `getCachedInfo`:`347` and `ReadJSON_Keys_Layout` comes from  `readLocalFile `: `367`.

Should I try to replace `MonadFail` with a `Throws _`?

Reply:
My `mtl` foo is not good enough to remove `MonadFail ReadJSON_Keys_Layout`. 

@edsko I see you wrote the `ReadJSON_Keys_Layout` code, do you have any advice?

Reply:
THe cloest I have gotten is `MonadError DeserializationError IO`, but this produces the warning:

```haskell
Security/Client.hs:211:18: warning:
    • Couldn't match type ‘IOException’ with ‘DeserializationError’
        arising from a functional dependency between:
          constraint ‘MonadError DeserializationError IO’
            arising from the type signature for:
                           checkForUpdates :: forall (down :: * -> *).
                                              (MonadError DeserializationError IO,
                                               Throws VerificationError, Throws SomeRemoteError) =>
                                              Repository down -> Maybe UTCTime -> IO HasUpdates
          instance ‘MonadError IOException IO’ at <no location info>
      Inaccessible code in
        a pattern with constructor: FGz :: Format FormatGz,
        in a case alternative
    • In the pattern: FGz
      In the pattern: Some FGz
      In a case alternative:
          Some FGz -> verifyFileInfo' (Just newInfoTarGz) targetPath tempPath
```